f.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2016 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+//
+// Modified by KEEP SEZC to expose F compression function.
+//
+package blake2
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math/bits"
+)
+
+const (
+	// BlockSize is the blocksize of BLAKE2 in bytes.
+	BlockSize = 128
+)
+
+var iv = [8]uint64{
+	0x6a09e667f3bcc908, 0xbb67ae8584caa73b, 0x3c6ef372fe94f82b, 0xa54ff53a5f1d36f1,
+	0x510e527fade682d1, 0x9b05688c2b3e6c1f, 0x1f83d9abfb41bd6b, 0x5be0cd19137e2179,
+}
+
+// the precomputed values for the first 12 rounds of BLAKE2
+// there are 12 16-byte arrays - one for each round
+// the entries are calculated from the sigma constants.
+var precomputed = [12][16]byte{
+	{0, 2, 4, 6, 1, 3, 5, 7, 8, 10, 12, 14, 9, 11, 13, 15},
+	{14, 4, 9, 13, 10, 8, 15, 6, 1, 0, 11, 5, 12, 2, 7, 3},
+	{11, 12, 5, 15, 8, 0, 2, 13, 10, 3, 7, 9, 14, 6, 1, 4},
+	{7, 3, 13, 11, 9, 1, 12, 14, 2, 5, 4, 15, 6, 10, 0, 8},
+	{9, 5, 2, 10, 0, 7, 4, 15, 14, 11, 6, 3, 1, 12, 8, 13},
+	{2, 6, 0, 8, 12, 10, 11, 3, 4, 7, 15, 1, 13, 5, 14, 9},
+	{12, 1, 14, 4, 5, 15, 13, 10, 0, 6, 9, 8, 7, 3, 2, 11},
+	{13, 7, 12, 3, 11, 14, 1, 9, 5, 15, 8, 2, 0, 4, 6, 10},
+	{6, 14, 11, 0, 15, 9, 3, 8, 12, 13, 1, 10, 2, 7, 4, 5},
+	{10, 8, 7, 1, 2, 4, 6, 5, 15, 9, 3, 13, 11, 14, 12, 0},
+	{0, 2, 4, 6, 1, 3, 5, 7, 8, 10, 12, 14, 9, 11, 13, 15}, // equal to the first
+	{14, 4, 9, 13, 10, 8, 15, 6, 1, 0, 11, 5, 12, 2, 7, 3}, // equal to the second
+}
+
+// F is a compression function for Blake2. It takes as an argument the state
+// vector `h`, message block vector `blocks` 2-bit offset counter `t`, final
+// block indicator flag `f`, and number of rounds `rounds`. The state vector
+// is modified by the function. Number of rounds can be anything between 0 and
+// 12 (inclusive).
+func F(h *[8]uint64, blocks []byte, t *[2]uint64, f uint64, rounds uint8) error {
+	if rounds > 12 {
+		return fmt.Errorf(
+			"number of rounds can not be greater than 12; is: [%v]",
+			rounds,
+		)
+	}
+
+	var m [16]uint64
+	t0, t1 := t[0], t[1]
+
+	for i := 0; i < len(blocks); {
+		t0 += BlockSize
+		if t0 < BlockSize {
+			t1++
+		}
+
+		v0, v1, v2, v3, v4, v5, v6, v7 := h[0], h[1], h[2], h[3], h[4], h[5], h[6], h[7]
+		v8, v9, v10, v11, v12, v13, v14, v15 := iv[0], iv[1], iv[2], iv[3], iv[4], iv[5], iv[6], iv[7]
+		v12 ^= t0
+		v13 ^= t1
+		v14 ^= f
+
+		for j := range m {
+			m[j] = binary.LittleEndian.Uint64(blocks[i:])
+			i += 8
+		}
+
+		for j := 0; j < int(rounds); j++ {
+			s := &(precomputed[j])
```

Comment:
We can support up to `2 ^ 8 - 1` rounds if we replace this `j` with `j mod 10` and get rid of the last two values in the `precompute` array. It'll be at a slight performance expense but 🤷‍♂️

Reply:
Done!